### ISSUE TITLE
Expose save states through the ui

### ISSUE THREAD

**Issue by andychase** [andychase/gbajs2#10]
This should work

```
<script>
function save() {
	let gba = this.gba;
	gba.pause();
	window.localforage.setItem("state", Serializer.serialize(gba.freeze()), function (err) {
		gba.runStable();
	})
}

function load() {
	let gba = this.gba;
	window.localforage.getItem("state", function (err, value) {
		Serializer.deserialize(value, function (out) {
			gba.pause();
			gba.defrost(out);
			gba.runStable();
		});
	});
}
</script>

<button class="bigbutton" onclick="save()">SAVE</button>
		<button class="bigbutton" onclick="load()">LOAD</button>
```

However we are running out of room in the ui, so some ui designs make be helpful

**Comment by andychase** [comment]
Also if that can be re-written without the use of localforage. The blob from save states needs to be encoded for localstorage

**Comment by andychase** [comment]
In one of my projects I tried to have the system save as someone was closing the window, but apparently you can't have any async/callbacks in the onclose handlers. Unfortunately I got stuck here because I could not figure out how to convert a blob to a string without using async which is needed for localStorage.

Although indexedDB works without the conversion, apparently won't save onclose either.

**Comment by andychase** [comment]
My solution was just save on interval. This caused a studder at first but I re-arranged the code above so only `gba.freeze()` runs while gba is paused. That function is fast enough not to cause a studder. Everything else can be done while the gba is unpaused

**Comment by Qaena** [comment]
Just thought I'd chime in to say that I also had this problem on a different project of mine that used local storage/db, and I couldn't find a way to do it either and also settled on an interval of sorts, so I'm not sure there's a better option.

**Comment by heavilyarmedporky** [comment]
Where do we insert this? Would this let me load download savegames (I asked on #23 )?: 
`<script>
function save() {
	let gba = this.gba;
	gba.pause();
	window.localforage.setItem("state", Serializer.serialize(gba.freeze()), function (err) {
		gba.runStable();
	})
}

function load() {
	let gba = this.gba;
	window.localforage.getItem("state", function (err, value) {
		Serializer.deserialize(value, function (out) {
			gba.pause();
			gba.defrost(out);
			gba.runStable();
		});
	});
}
</script>

<button class="bigbutton" onclick="save()">SAVE</button>
		<button class="bigbutton" onclick="load()">LOAD</button>`

**Comment by andychase** [comment]
Yeah I think it would let you download/upload save games. And it could go in index.html, or somewhere else.

If you are going to look at this, it might be worth taking a look at this commit I made which fixed it from the original source:

https://github.com/andychase/gbajs2/commit/50a88fe7a5ed3ccf58e31e07e65d27723b7885fb

The GameBoyAdvanceRenderProxy() was the renderer the original project used, I think the intention was to use service workers as background threads to get higher performance, however in my testing I didn't see a difference and I think the service workers makes things more complicated. The render proxy didn't have save states implemented so that's why I swapped back.

Also save states were completely broken before that register change which I'm kind of shocked I was able to debug I kind of stumbled on it.

Another thing to note on save states is while loading, I have noticed that sometimes the screen/sound is glitchy for a few frames before just suddenly working fine. This might have to do with video/audio registers, but either way it didn't matter for what I was trying to do so I didn't look into it.

**Comment by absolute00** [comment]
what file to put the code in, and where in the code?

**Comment by DragonMaster12345678** [comment]
Should I be getting a prompt when I click on the save button added through the first comment? If so, is there any additional code I need to import to make it run properly? Also when I add the code in the first comment to index.html it causes the screen to go off center? 

![image](https://github.com/andychase/gbajs2/assets/95725497/2f90b3d5-4236-4171-a37c-41db71d26797)

**Comment by andychase** [comment]
I had made this change, not sure maybe it's a buggy change

https://github.com/andychase/gbajs2/commit/50a88fe7a5ed3ccf58e31e07e65d27723b7885fb
